src/panels/lovelace/editor/card-editor/hui-card-editor.ts

```diff
@@ -0,0 +1,264 @@
+    };
+    "config-changed": {
+      config: LovelaceCardConfig;
+    };
+  }
+}
+
+@customElement("hui-card-editor")
+export class HuiCardEditor extends LitElement {
+  @property() public hass?: HomeAssistant;
+  @property() public value?;
+
+  @property() private _yaml?: string | undefined;
+  @property() private _configElement?: LovelaceCardEditor;
+  @property() private _configElType?: string;
+  @property() private _GUImode: boolean = true;
+  // Error: Configuration broken - do not save
+  @property() private _error?: string;
+  // Warning: GUI editor can't handle configuration - ok to save
+  @property() private _warning?: string;
+  @property() private _loading: boolean = false;
+
+  public get yaml(): string {
+    return this._yaml || "";
+  }
+  public set yaml(_yaml: string) {
+    this._yaml = _yaml;
+    this._updateConfigElement();
+    // TODO: Defer this to next redraw? How?
+    setTimeout(() => this._yamlEditor.codemirror.refresh(), 1);
+    fireEvent(this, "config-changed", { config: this.config! });
+  }
+
+  public get config(): LovelaceCardConfig | undefined {
+    try {
+      return yaml.safeLoad(this.yaml);
+    } catch (err) {
+      this._error = err;
+      return { type: "error", error: err.message };
+    }
+  }
+  public set config(config: LovelaceCardConfig | undefined) {
+    this.yaml = yaml.safeDump(config);
+  }
+
+  public get error(): boolean {
+    return this._error !== undefined;
+  }
+
+  private get _yamlEditor(): HuiYamlEditor {
+    return this.shadowRoot!.querySelector("hui-yaml-editor")!;
+  }
+
+  protected render(): TemplateResult {
+    return html`
+      <div style="width: 100%;">
+        <div class="gui-editor" ?hidden="${this._GUImode === false}">
+          ${this._loading
+            ? html`
+                <paper-spinner
+                  ?active="${this._loading}"
+                  alt="Loading"
+                  class="center margin-bot"
+                ></paper-spinner>
+              `
+            : this._warning || this._error
+            ? ""
+            : this._configElement}
+        </div>
+        <div class="yaml-editor" ?hidden="${this._GUImode === true}">
+          <hui-yaml-editor
+            .hass="${this.hass}"
+            .value="${this.yaml}"
+            @yaml-changed="${this._handleYAMLChanged}"
+          ></hui-yaml-editor>
+          <div class="error">
+            ${this._error}
+          </div>
+          <div class="warning">
+            ${this._warning}
+          </div>
+        </div>
+        <div class="buttons">
+          <mwc-button
+            @click="${() => (this._GUImode = !this._GUImode)}"
+            ?disabled="${this._warning || this._error}"
+            ?unelevated="${this._GUImode === false}"
+          >
+            <ha-icon .icon="${"mdi:code-braces"}"></ha-icon>
+          </mwc-button>
+        </div>
+      </div>
+    `;
+  }
+
+  static get styles(): CSSResult {
+    return css`
+      :host {
+        margin: 0 10px;
```

Comment:
We should never add margin to our elements. Margin has to be applied by whoever element embeds this. We don't know how this will be used, and so can't decide that a margin is always necessary. If a margin is necessary for rendering, we should accomplish that with a padding.